sm_ipt/sm_ipt.c

```diff
@@ -0,0 +1,420 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+
+#include "sm_ipt.h"
+#include "sm_ipt_log.h"
+#include "sm_ipt_errno.h"
+
+
+#define FLAG_RELEASE 0x80
+
+#define WORD_SIZE sizeof(uint32_t)
+#define WORD_BITS (8 * sizeof(uint32_t))
+
+#if defined(CONFIG_SM_IPT_NUM_BLOCKS_32)
+typedef uint32_t mask_t;
+typedef int32_t smask_t;
+#define NUM_BLOCKS 32
+#define mask_clz sm_ipt_os_clz32
+#elif defined(CONFIG_SM_IPT_NUM_BLOCKS_64)
+typedef uint64_t mask_t;
+typedef int64_t smask_t;
+#define NUM_BLOCKS 64
+#define mask_clz sm_ipt_os_clz64
+#else
+#error Number of shared dynamic memory blocks is not configured.
+#endif
+
+#define ALLOCABLE_MULTIPLY (WORD_SIZE * NUM_BLOCKS)
+#define QUEUE_INDEX_MASK (NUM_BLOCKS - 1)
+
+#define memory_corrupted_error() sm_ipt_os_fatal(); \
+	SM_IPT_ERR("Shared memory corrupted"); \
+	SM_IPT_ASSERT(0)
+
+
+static inline void free_mask_set(struct sm_ipt_ctx *ctx, mask_t mask)
+{
+	sm_ipt_os_atomic_or(&ctx->free_mask[0], mask);
+	if (NUM_BLOCKS > 32) {
+		sm_ipt_os_atomic_or(&ctx->free_mask[1], (uint64_t)mask >> 32);
+	}
+	SM_IPT_OS_MEMORY_BARIER();
+}
+
+static bool free_mask_unset(struct sm_ipt_ctx *ctx, mask_t mask)
+{
+	uint32_t old;
+	uint32_t mask_part;
+
+	mask_part = (uint32_t)mask;
+	old = sm_ipt_os_atomic_and(&ctx->free_mask[0], ~mask_part);
+	if ((old & mask_part) != mask_part) {
+		sm_ipt_os_atomic_or(&ctx->free_mask[0], old & mask_part);
```

Comment:
Is this some kind of fallback mechanism ?

Reply:
Yes, if operation atomic_and was not done as expected - restore changes.

Reply:
... and return false.

This is needed, because I avoided locks, so in some rare cases, other thread may change `free_mask` after this thread calculated the `mask` and before `atomic_and`. In such case whole operation in repeated in current thread.